File: algorithm/minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    var map = {};
    for (var i = 0; i < t.length; i++) {
        if (map[t[i]]) {
            map[t[i]]++;
        } else {
            map[t[i]] = 1;
        }
    }
    var left = 0, right = 0, minLen = Number.MAX_VALUE, minLeft = 0, minRight = 0;
    while (right < s.length) {
        if (map[s[right]]) {
            map[s[right]]--;
            if (map[s[right]] >= 0) {
                while (map[s[left]] < 0) {
                    map[s[left]]++;
                    left++;
                }
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                    minRight = right;
                }
            }
        }
        right++;
    }
    return minLen === Number.MAX_VALUE ? '' : s.substring(minLeft, minRight + 1);    
};

File: algorithm/minimum-window-substring.c
char * minWindow(char * s, char * t){
    int i, j, k, l, m;
    int count = 0;
    int len = 0;
    int left = 0;
    int right = 0;
    char *str = (char *)malloc(sizeof(char) * 100);
    for (i = 0; i < strlen(s); i++) {
        for (j = 0; j < strlen(t); j++) {
            if (s[i] == t[j]) {
                str[len++] = s[i];
                left++;
            } else {
                if (left > right) {
                    str[len++] = s[i];
                    left++;
                } else {
                    str[len++] = s[i];
                    right++;
                }
            }
        }
        str[len++] = s[i];
        right++;
        for (k = 0; k < len; k++) {
            if (str[k] == '(') {
                left++;
            } else {
                right--;
            }
            if (left == right) {
                count++;
            }
        }
        len = 0;
        left = 0;
        right = 0;
    }
    return str;
}

// Language: c
// Path: algorithm\generate-parentheses.c